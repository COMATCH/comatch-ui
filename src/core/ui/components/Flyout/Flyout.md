## Flyout

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**button** | `Union<Function\|ReactNode>` | `'Flyout'` | :x: | 
**button<1>** | `Function` |  | :x: | 
**button<2>** | `ReactNode` |  | :x: | 
**children** | `Union<Array[]<ReactNode>\|ReactNode>` |  | :white_check_mark: | 
**children<1>** | `Array[]<ReactNode>` |  | :x: | 
**children<2>** | `ReactNode` |  | :x: | 
**className** | `String` | `''` | :x: | 
**closeOnClickOutside** | `Boolean` | `true` | :x: | 
**delayMs** | `Number` | `250` | :x: | 
**isOpen** | `Boolean` | `false` | :x: | 
**onClickOutside** | `Function` | `noop` | :x: | 
**toggle** | `Boolean` | `true` | :x: | 

