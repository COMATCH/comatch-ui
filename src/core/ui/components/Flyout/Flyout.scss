@import '../../../styles/variables.scss';
@import '../../../styles/mixins.scss';

.Flyout {
    $flyout: &;
    $defaultTopPosition: 15%;

    display: flex;
    position: fixed;
    right: 0;
    top: $defaultTopPosition;
    z-index: 1;

    &__modal {
        background: rgba(62, 62, 62, 0.8);
        height: 100%;
        opacity: 0;
        padding: 5px;
        transition: opacity 500ms ease-in;

        &--fade-in {
            opacity: 1;
        }

        &--fade-out {
            opacity: 0;
        }

        @include desktop {
            background: none;
            padding: 0;
        }
    }

    &__content {
        background: $co-color-white;
        border-radius: 3px;
        box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.5);
        padding: 15px;

        @include desktop {
            width: 400px;
        }
    }

    &__button-wrapper {
        margin: 50px 0 0 10px;
    }

    &--open {
        flex-direction: column;
        height: 100%;
        justify-content: center;
        top: 0;
        width: 100%;

        #{$flyout}__button-wrapper {
            display: none;
        }

        @include desktop {
            height: auto;
            top: $defaultTopPosition;
            width: auto;
            flex-direction: row;
            background: none;
            padding: 0;

            #{$flyout}__button-wrapper {
                display: block;
            }
        }
    }
}
