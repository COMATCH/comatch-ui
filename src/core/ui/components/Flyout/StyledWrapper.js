import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { device } from '../../../styles/mixins';
import { palette } from '../../../styles/variables';

const ROOT = 'Flyout';
const MODAL = `${ROOT}__modal`;
const MODAL_FADE_IN = `${MODAL}--fade-in`;
const MODAL_FADE_OUT = `${MODAL}--fade-out`;

const StyledModalContent = styled.div`
    background: ${palette.white};
    border-radius: 3px;
    box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.5);
    padding: 15px;

    @media ${device.desktop} {
        width: 400px;
    }
`;

const StyledModal = styled.div`
    background: rgba(62, 62, 62, 0.8);
    height: 100%;
    opacity: 0;
    padding: 5px;
    transition: opacity 500ms ease-in;

    &.${MODAL_FADE_IN} {
        opacity: 1;
    }

    &.${MODAL_FADE_OUT} {
        opacity: 0;
    }

    @media ${device.desktop} {
        background: none;
        padding: 0;
    }
`;

const StyledButtonWrapper = styled.div`
    margin: 50px 0 0 10px;
`;

const StyledWrapper = styled(forwardRef(({ open, ...rest }, ref) => <div {...rest} ref={ref} />))`
    display: flex;
    position: fixed;
    right: 0;
    top: 15%;
    z-index: 1;

    ${({ open }) => {
        if (!open) {
            return '';
        }

        return `
            flex-direction: column;
            height: 100%;
            justify-content: center;
            top: 0;
            width: 100%;

            ${StyledButtonWrapper} {
                display: none;
            }

            @media ${device.desktop} {
                height: auto;
                top: 15%;
                width: auto;
                flex-direction: row;
                background: none;
                padding: 0;

                ${StyledButtonWrapper} {
                    display: block;
                }
            }
        `;
    }}
`;

export { StyledButtonWrapper, StyledModal, StyledModalContent, StyledWrapper };
