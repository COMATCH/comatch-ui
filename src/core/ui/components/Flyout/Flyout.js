import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import isFunction from 'lodash/isFunction';

import { StyledButtonWrapper, StyledContentWrapper, StyledContent, StyledWrapper } from './StyledWrapper';

/*
 * `onTargetClick` triggers only an event if the
 * event comes from its actual target and not from
 * a child. This can be used to prevent events being
 * triggered by childs through event bubbling.
 */
const onTargetClick = (click) => (event) => {
    if (event.target === event.currentTarget) {
        click(event);
    }
};

export class Flyout extends Component {
    static propTypes = {
        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
        className: PropTypes.string,
        delayMs: PropTypes.number,
        isOpen: PropTypes.bool,
        button: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    };

    static defaultProps = {
        className: '',
        delayMs: 250,
        isOpen: false,
        button: 'Flyout',
    };

    state = {
        isOpen: this.props.isOpen,
        fadeout: !this.props.isOpen,
    };

    componentWillUnmount() {
        if (this.timerId) {
            this.clearTimeout();
        }
    }

    clearTimeout = () => {
        if (this.timerId) {
            clearTimeout(this.timerId);
            this.timerId = undefined;
        }
    };

    delay = (fn, delay) => {
        this.clearTimeout();
        this.timerId = setTimeout(fn, delay);
    };

    open = () => {
        this.setState({ isOpen: true, fadeout: true });
        this.delay(() => this.setState({ isOpen: true, fadeout: false }), 0);
    };

    close = () => {
        this.setState({ isOpen: true, fadeout: true });
        this.delay(() => this.setState({ isOpen: false, fadeout: true }), this.props.delayMs);
    };

    toggleOpen = () => {
        if (this.state.isOpen) {
            this.close();
        } else {
            this.open();
        }
    };

    render() {
        const { toggleOpen } = this;
        const { children, className, button, delayMs } = this.props;
        const { isOpen, fadeout } = this.state;
        const fadeTransition = {
            transition: `opacity ${delayMs}ms ease-in`,
            opacity: fadeout ? 0 : 1,
        };

        return (
            <StyledWrapper open={isOpen} className={classNames('Flyout', className, { 'Flyout--open': isOpen })}>
                {isOpen && (
                    <StyledContentWrapper
                        onClick={onTargetClick(toggleOpen)}
                        className="Flyout__content-wrapper"
                        style={fadeTransition}
                    >
                        <StyledContent className="Flyout__content">{children}</StyledContent>
                    </StyledContentWrapper>
                )}
                <StyledButtonWrapper className="Flyout__button-wrapper" onClick={toggleOpen}>
                    {isFunction(button) ? button() : button}
                </StyledButtonWrapper>
            </StyledWrapper>
        );
    }
}
