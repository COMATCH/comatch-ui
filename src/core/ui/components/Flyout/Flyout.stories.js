import React, { Component } from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { Button } from '../Button/Button';
import { Flyout } from './Flyout';

const LoremText = () => (
    <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed et accumsan libero, ac varius turpis. Quisque
        euismod, dolor sed finibus sollicitudin, lorem risus pulvinar neque, quis ultrices sapien elit nec odio. Morbi
        elementum purus sapien, at aliquam tortor volutpat ut. Sed nisl massa, dapibus at suscipit et, sollicitudin
        vitae orci. Vestibulum faucibus lacus vitae urna porta, quis efficitur massa semper. Duis feugiat blandit ligula
        nec sodales. Sed ultricies vitae nunc eu tempus. Sed vitae accumsan ipsum. Nunc eu pulvinar lacus. Ut ac
        tincidunt arcu. Aliquam varius congue mi in placerat. Vestibulum quam quam, malesuada a ipsum et, dictum
        scelerisque ante. Donec lobortis nibh eget iaculis vehicula. Ut ornare interdum lobortis.
    </div>
);

class FlyoutWrapper extends Component {
    render() {
        return (
            <Flyout
                ref={(ref) => {
                    this.flyout = ref;
                }}
            >
                <LoremText />
                <Button onClick={() => this.flyout.close()} text="Close" />
            </Flyout>
        );
    }
}

storiesOf('Flyout', module)
    .add('show flyout on page', () => (
        <>
            <Flyout>
                <LoremText />
            </Flyout>
            <Highlight className="html">{'<Flyout> <div>Lorem ipsum...</div> </Flyout>'}</Highlight>
        </>
    ))
    .add('close with button placed in the flyout', () => (
        <>
            <FlyoutWrapper />
            <Highlight className="html">
                {'<Flyout ref={(ref) => { this.flyout = ref;}}> <div>Lorem ipsum...</div> ' +
                    '<Button onClick={() => this.flyout.close()} text="Close" /> </Flyout>'}
            </Highlight>
        </>
    ));
