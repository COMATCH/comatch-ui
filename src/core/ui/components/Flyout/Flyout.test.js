import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { lorem } from 'faker';
import { Flyout } from './Flyout';
import '../../../config/tests/setup';

describe('Flyout', () => {
    it('should render correctly', () => {
        const flyout = renderer.create(
            <Flyout>
                <div>
                    some-content
                </div>
            </Flyout>
        ).toJSON();

        expect(flyout).toMatchSnapshot();
    });

    it('should render with button content correctly', () => {
        const flyout = renderer.create(
            <Flyout
                button={
                    <span>sample-button-text</span>
                }
            >
                <div>
                    some-content
                </div>
            </Flyout>
        ).toJSON();

        expect(flyout).toMatchSnapshot();
    });

    // jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;
    it('should open the flyout after click on the button', () => {

        // fake the timers inside the component
        jest.useFakeTimers();

        const wrapper = mount(
            <Flyout>
                <div>
                    some-content
                </div>
            </Flyout>
        );

        const button = wrapper.find('div.Flyout__button-wrapper');

        button.simulate('click');
        let flyoutContent = wrapper.find('div.Flyout__content-wrapper');
        expect(flyoutContent.length).toEqual(1);

        button.simulate('click');

        jest.runAllTimers();
        wrapper.update();

        flyoutContent = wrapper.find('div.Flyout__content-wrapper');
        expect(flyoutContent.length).toEqual(0);
    });

    it('should close the flyout when clicking on content wrapper itself', () => {

        const className = lorem.word();

        // fake the timers inside the component
        jest.useFakeTimers();

        const wrapper = mount(
            <Flyout>
                <div className={className}>
                    some-content
                </div>
            </Flyout>
        );

        // open the flyout, no need to assert here, already covered before.
        const button = wrapper.find('div.Flyout__button-wrapper');
        button.simulate('click');
        let flyoutContent = wrapper.find('div.Flyout__content-wrapper');

        // if the click was on a child elem, the close event should be ignored
        const content = wrapper.find(`div.${className}`);
        content.simulate('click');
        jest.runAllTimers();
        wrapper.update();

        flyoutContent = wrapper.find('div.Flyout__content-wrapper');
        expect(flyoutContent.length).toEqual(1);

        // now if the click on on the wrap itself, the flyout should be closed
        const contentWrap = wrapper.find('div.Flyout__content-wrapper');
        contentWrap.simulate('click');
        jest.runAllTimers();
        wrapper.update();

        flyoutContent = wrapper.find('div.Flyout__content-wrapper');
        expect(flyoutContent.length).toEqual(0);
    });
});
