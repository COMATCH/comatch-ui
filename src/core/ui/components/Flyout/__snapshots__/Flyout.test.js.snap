// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flyout should render correctly 1`] = `
<div
  className="Flyout sc-bxivhb jrcNjb"
>
  <div
    className="Flyout__button-wrapper sc-htpNat cSppWP"
    onClick={[Function]}
  >
    Flyout
  </div>
</div>
`;

exports[`Flyout should render with button content correctly 1`] = `
<div
  className="Flyout sc-bxivhb jrcNjb"
>
  <div
    className="Flyout__button-wrapper sc-htpNat cSppWP"
    onClick={[Function]}
  >
    <span>
      sample-button-text
    </span>
  </div>
</div>
`;
